### PR TITLE
[WIP]Added deribit.

### PR DESCRIPTION
**Pull request by leofisG**
market:
- GetOrderBook
- GetInstrument
- GetCurrencies

user:
- buy
- sell
- cancel
- cancelAll
- cancelByLabel
- getOrderState

### CODE REVIEW

File: deribit-api/src/main/java/io/contek/invoker/deribit/api/websocket/WebSocketApi.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  }
+
+  @Override
+  protected WebSocketCall createCall(ICredential credential) {
+    return WebSocketCall.fromUrl(context.getBaseUrl() + "/ws/api/v2");
+  }
+
+  @Override
+  protected final void checkErrorMessage(AnyWebSocketMessage message) {
+    if (message instanceof WebSocketInfoMessage) {
```
Comment (beijunyi) [inline deribit-api/src/main/java/io/contek/invoker/deribit/api/websocket/WebSocketApi.java:34]: I believe deribit does not have such message type. Delete this class and remove any logic corresponds to this type.

File: deribit-api/src/main/java/io/contek/invoker/deribit/api/websocket/WebSocketChannel.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  @Override
+  protected final SubscriptionState subscribe(WebSocketSession session) {
+
+    synchronized (pendingSubscriptionHolder) {
+      if (pendingSubscriptionHolder.get() != null) {
+        throw new IllegalStateException();
+      }
+      WebSocketRequest request = new WebSocketRequest();
+      request.method = "public/" + WebSocketOutboundKeys._subscribe;
+      request.params.put("channels", Collections.singletonList(getChannel()));
```
Comment (beijunyi) [inline deribit-api/src/main/java/io/contek/invoker/deribit/api/websocket/WebSocketChannel.java:36]: Always use Immutable(List/Set/Map) to be more consistent with the rest of the code base.

File: deribit-api/src/main/java/io/contek/invoker/deribit/api/websocket/WebSocketChannel.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+  @Override
+  protected final SubscriptionState unsubscribe(WebSocketSession session) {
+    synchronized (pendingSubscriptionHolder) {
+      if (pendingSubscriptionHolder.get() != null) {
+        throw new IllegalStateException();
+      }
+      WebSocketRequest request = new WebSocketRequest();
+      request.method = "public/" + WebSocketOutboundKeys._unsubscribe;
+      request.params.put("channels", Collections.singletonList(getChannel()));
```
Comment (beijunyi) [inline deribit-api/src/main/java/io/contek/invoker/deribit/api/websocket/WebSocketChannel.java:52]: same here

File: ftx-api/src/main/java/io/contek/invoker/ftx/api/ApiFactory.java [hunk 4]
```diff
@@ -28,6 +29,13 @@
 @ThreadSafe
 public final class ApiFactory {
 
+  public static void main(String[] args) {
```
Comment (beijunyi) [inline ftx-api/src/main/java/io/contek/invoker/ftx/api/ApiFactory.java:32]: Delete test purpose code

File: deribit-api/src/main/java/io/contek/invoker/deribit/api/websocket/market/TradesChannel.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  protected Class<TradesChannel.Message> getMessageType() {
+    return TradesChannel.Message.class;
+  }
+
+  @Override
+  protected boolean accepts(TradesChannel.Message message) {
+    return instrumentName.equals(parseInstrumentName(message.params.channel));
+  }
+
+  private String parseInstrumentName(String name) {
```
Comment (beijunyi) [inline deribit-api/src/main/java/io/contek/invoker/deribit/api/websocket/market/TradesChannel.java:45]: static

File: deribit-api/src/main/java/io/contek/invoker/deribit/api/websocket/market/TradesChannel.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+import javax.annotation.concurrent.NotThreadSafe;
+import javax.annotation.concurrent.ThreadSafe;
+import java.util.List;
+
+import static io.contek.invoker.deribit.api.websocket.common.constants.WebSocketChannelKeys._trades;
+
+@ThreadSafe
+public final class TradesChannel extends WebSocketChannel<TradesChannel.Message> {
+
+  private final String instrumentName;
```
Comment (beijunyi) [inline deribit-api/src/main/java/io/contek/invoker/deribit/api/websocket/market/TradesChannel.java:19]: the full channel name should be kept here.
and then we don't have to extract the instrument name when we determine whether we should accept a message.

File: deribit-api/src/main/java/io/contek/invoker/deribit/api/websocket/market/TradesChannel.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+  }
+
+  @Override
+  protected String getChannel() {
+    return String.format("%s.%s.%s", _trades, instrumentName, "100ms");
+  }
+
+  @Override
+  protected String getDisplayName() {
+    return Joiner.on(':').join(WebSocketChannelKeys._trades, instrumentName);
```
Comment (beijunyi) [inline deribit-api/src/main/java/io/contek/invoker/deribit/api/websocket/market/TradesChannel.java:32]: If we keep the whole channel name. getChannel() and getDisplayName() can be simplified,

File: deribit-api/src/main/java/io/contek/invoker/deribit/api/websocket/market/OrderBookChannel.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+import javax.annotation.concurrent.NotThreadSafe;
+import javax.annotation.concurrent.ThreadSafe;
+
+import static io.contek.invoker.deribit.api.websocket.common.constants.WebSocketChannelKeys._orderbook;
+
+@ThreadSafe
+public final class OrderBookChannel extends WebSocketChannel<OrderBookChannel.Message> {
+
+  private final String instrumentName;
```
Comment (beijunyi) [inline deribit-api/src/main/java/io/contek/invoker/deribit/api/websocket/market/OrderBookChannel.java:17]: See the comments in TradesChannel.

File: deribit-api/src/main/java/io/contek/invoker/deribit/api/websocket/common/WebSocketRequest.java [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+  public String method;
+  public Integer id;
+  public String jsonrpc = "2.0";
+  public Map<String, List<String>> params;
+  private static int nextId = 0;
+
+  public WebSocketRequest() {
+    this.params = new HashMap<>();
+    this.id = nextId;
+    nextId += 1;
```
Comment (beijunyi) [inline deribit-api/src/main/java/io/contek/invoker/deribit/api/websocket/common/WebSocketRequest.java:22]: This class is a bit hackery. But let's leave it like this for now.
This class will work until we want to use websocket for more than realtime data subscription

File: deribit-api/src/main/java/io/contek/invoker/deribit/api/websocket/common/WebSocketInfoMessage.java [hunk 10]
```diff
@@ -0,0 +1,10 @@
+package io.contek.invoker.deribit.api.websocket.common;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+@NotThreadSafe
+public final class WebSocketInfoMessage extends WebSocketInboundMessage {
```
Comment (beijunyi) [inline deribit-api/src/main/java/io/contek/invoker/deribit/api/websocket/common/WebSocketInfoMessage.java:6]: Delete this class.

File: deribit-api/src/main/java/io/contek/invoker/deribit/api/rest/RestRequest.java [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+
+  private ImmutableMap<String, String> generateHeaders(
+    RestMethod method, String paramsString, String bodyString, ICredential credential) {
+
+    if (credential.isAnonymous()) {
+      return ImmutableMap.of();
+    }
+    String clientId = credential.getApiKeyId();
+    String timestamp = String.valueOf(clock.millis());
+    String nonce = UUID.randomUUID().toString().substring(0, 8);
```
Comment (beijunyi) [inline deribit-api/src/main/java/io/contek/invoker/deribit/api/rest/RestRequest.java:77]: This is OK, but don't use UUID when you don't need a UUID. UUID is heavier than other ways to generate random string.
You can use something like BaseEncoding.base32().encode(Random.nextBytes(8))) to generate nonce.

File: deribit-api/src/main/java/io/contek/invoker/deribit/api/common/_Instrument.java [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+  public double maker_commission;
+  public double min_trade_amount;
+  public String option_type;
+  public String quote_currency;
+  public String settlement_period;
+  public double strike;
+  public double taker_commission;
+  public double tick_size;
+
+  @Override
```
Comment (beijunyi) [inline deribit-api/src/main/java/io/contek/invoker/deribit/api/common/_Instrument.java:26]: Delete toString as we shouldn't need to print class types.

**Review by pimpcapital — CHANGES_REQUESTED** [review]
Very well done.
Please see the comments for the minor refactoring works.

**Pull request merged**